pennylane/_qubit_device.py

```diff
@@ -688,3 +689,107 @@ def sample(self, observable):
         unraveled_indices = [2] * len(device_wires)
         indices = np.ravel_multi_index(samples.T, unraveled_indices)
         return observable.eigvals[indices]
+
+    def rewind_jacobian(self, tape):
+        """Implements the method outlined in https://arxiv.org/abs/2009.02823 to calculate the
+        Jacobian."""
+
+        # Perform the forward pass
+        self.reset()
+        self.execute(tape)
```

Comment:
It would be great to avoid this if possible. For example, consider autograd:

```python
>>> grad_fn = qml.grad(qnode)
>>> grad_fn(weights)  # executes both the forward and backward pass
```

In the case of autograd, we know that there will _always_ be a call to `execute` directly prior to the call to `jacobian()`, so this extra execute call here is wasteful - we are performing a redundant device execution.

This is not always the case in Torch/TF, since they separate the forward and backward calls in the UI:

```python
>>> loss = qnode(weights)  # forward pass
>>> qnode(new_weights) # another forward pass
>>> loss.backward() # backward pass
```

It would be great if we could determine if the _last execution_ used the same tape and the same parameter values, and somehow avoid a redundant execution. E.g.,

```python
forward_pass_exists = tape is self._last_tape and np.allclose(tape.get_parameters(), self._last_parameters)

if not forward_pass_exists:
    self.reset()
    self.execute(tape)
```